### DISCUSSION TITLE
Is not reporting on covariant TypeVars in Protocols when using it for a private variable intentional?

### DISCUSSION THREAD

**Discussion by Daraan** [microsoft/pyright#8841] [Q&A] [answered]
The situation is like this:

```py
T = TypeVar("T", covariant=True)

# Type variable "T" used in generic protocol "P1" should be invariant
class P1(Protocol[T]):
    value: T

# OK
class P2(Protocol[T]):
    _value: T
```

Is it intentional that for `P2` no errors are reported; mypy reports both? If yes, could you give me some insights into the idea behind the decision?

Related discussion: https://github.com/microsoft/pyright/discussions/3090

**Comment by erictraut** [comment, marked as answer]
I added an exemption for variables with protected or private names based on [this discussion](https://github.com/microsoft/pyright/discussions/5568#discussioncomment-6582748). This also affects auto-invariance calculations for TypeVars that use the new (PEP 695) syntax. 

In hindsight, I wonder if this was the right decision — at least for protected variables. I think it's justified for private variables because they cannot be modified outside of the class and cannot be overridden by subclasses (because of name mangling), but I'm not as sure about variables with protected names.

The Python typing spec is silent on this topic. Because this is a gray area, I'm reluctant to change pyright's behavior unless/until there is consensus in the community about the "correct" behavior. If you feel strongly about it, you're welcome to start a discussion in the [Python typing forum](https://discuss.python.org/c/typing/32).

**Reply by Daraan** [reply]
@erictraut Thank you for the reply.

I now ran into this again from the other side: I want an invariant protected field but pyright suggest I should use a covariant on it.
First I would assume this is okay, secondly from the linked discussion #5568 I would assume that they are excluded from the checking.
Now I am confused why does pyright suggest it to be the opposite way for a protected type variable? 

Code sample in [pyright playground](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgGoCGIANFAArgxgDGYANgFDsFQC8hJFEAAoARAREBKdgH0C0hCCQA3cjGK9%2BCMpVGz5ilWsky5C5auLSWGolsG7TBi1bAiaLFYvIoYPAiABXYik5az5bbWExFzcoD0okb19-IKl2AGJNdU9EgCNWdRE9M0NiESgAgGdiABNMFCg0YhRiRSY6BmY2KBEAMTBXKEqACzAA1jrc9XivHwysqBzyfMLQwara%2Bsbm1qR2%2BjBGFlYe-sGRsYmoKfqluaZWckrKqDOhA6O2AG0CGmKnNR-RzmNQuGhrAC6EgAXOwoPCoPoQZZUEpoYjgaUEZkADwAWjxQ1G40m0zAdxgAG4oMQAB5aJhqOoAeQA0nCEQgAvk9tJUejuAj4Zk2ZwhUjSi50f9kS5hVA2dS6cRsDAOfCuTymFLCHL5RcSdd1KiEklqU8afTiIzakA)

```python
from typing import TypeVar, Protocol

T = TypeVar("T")
_T_private = TypeVar("_T_private")
_T_private_co = TypeVar("_T_private_co", covariant=True)
T_co = TypeVar("T_co", covariant=True)

# Type variable "_T_private" used in generic Protocol "Foo" should be covariant
# Type variable "T_co" used in generic Protocol "Foo" should be invariant
class Foo(Protocol[T, _T_private, _T_private_co, T_co]):
    _private_inv: _T_private   # <-- should be covariant; expected OK
    public_inv: T      # OK

    _private_co: _T_private_co  # OK; exempt
    public_co: T_co   # should be invariant; as expected
```

**Reply by Daraan** [reply]
small ping :)

Could you help me with the confusing part in my last comment
